### ISSUE TITLE
imp's SafeTensors path is 41% off on a GDN hybrid — proven against an HF reference, and it confounds #1273

### ISSUE THREAD

**Issue by kekzl** [kekzl/imp#1287]
Found while re-checking the #1273 claim that "every hybrid NVFP4 checkpoint degrades". That framing may be wrong: **the degradation is still there when no NVFP4 is involved at all.**

## The measurement

Same engine, same corpus, same sitting, GPU exclusive, `build-dev`. Qwen3.5-4B has a true twin here: `/mnt/data/models/Qwen3.5-4B-BF16` is `Qwen/Qwen3.5-4B`, and unsloth's `Qwen3.5-4B-mxfp4.gguf` carries `general.base_model.0.repo_url = https://huggingface.co/Qwen/Qwen3.5-4B`. Both load as `layers=32 d_model=2560 d_ff=9216 heads=16 kv_heads=4 vocab=248320`, 8 attention / 24 GDN, and both tokenize the corpus to the same 13811 tokens.

| model | attention | SafeTensors | GGUF | ordering |
|---|---|---|---|---|
| Qwen3-14B | dense | **5.3769** (BF16) | 5.4545 (Q6_K) | BF16 better ✅ |
| Qwen3.5-4B | **hybrid** | 6.6238 (BF16) | **5.0928** (mxfp4) | BF16 worse ❌ |

`ppl_corpus.txt` (199 tokens) for both rows so the two are comparable. On `ppl_corpus_45k.txt` the hybrid row is the same story with more margin:

| arm | runs | spread |
|---|---|---|
| Qwen3.5-4B BF16 SafeTensors | 12.6794 / 12.6794 / 12.6798 | 0.003% |
| Qwen3.5-4B mxfp4 GGUF | 9.3869 / 9.3793 / 9.3948 | 0.17% |

Interleaved, one sitting. **A 4-bit quantization cannot beat its own 16-bit original by 26%** — 4 bits add no information. One of the two paths is wrong, and it is not the one with less precision.

## Why this matters for #1273

Every degraded checkpoint in #1273 is a **hybrid loaded from SafeTensors**; every healthy twin it is compared against is a **GGUF**. That confound was never controlled, because until #1282 no BF16 hybrid on this box produced a valid perplexity number at all (uniform logits, PPL = vocab_size).

Now that one does, the degradation is visible **without any NVFP4**, at BF16, on the same base model. So "NVFP4 degrades hybrids" is not established by the data in #1273; "imp's SafeTensors path degrades hybrids" fits every row of it, including why dense NVFP4 checkpoints are fine.

Our own NVFP4 export of this model, quantized from that same BF16 SafeTensors checkpoint, scores 14.88 — worse again, in the same direction. Everything that goes through the SafeTensors path on this model is ~30% off the GGUF path.

## What is already ruled out

| candidate | why |
|---|---|
| the SafeTensors loader in general | the dense 14B row is correctly ordered, same loader, same corpus |
| a tokenizer mismatch | both paths tokenize the corpus to the same 13811 tokens |
| run-to-run noise | spreads 0.003% and 0.17% against a 35% gap |
| a different base model | GGUF metadata names `Qwen/Qwen3.5-4B`; geometry, layer census and vocab all match |
| the `A_log` transform (#1282) | already fixed; that bug produced uniform logits, this one produces coherent-but-worse output |

## Not yet localized

A per-layer hidden-state diff between the two paths (`diagnostics.dump_hidden_dir=all`, prefill, identical tokens) shows relative divergence **flat at 0.17–0.22 across all 32 layers**, starting at 0.2558 on the embedding output itself. No step change at any layer, GDN or attention. That is consistent with the comparison being dominated by the 4-bit weights on the GGUF side, so this diff cannot say which path is wrong — it only says the error is not one broken layer in the middle.

The likely next step is the method that found #1282: compare the two loaders **tensor by tensor** on the GDN parameters, looking for another transform that one path applies and the other does not.

## Limitations, stated

- The precision differs on the hybrid row (BF16 vs mxfp4). That weakens nothing — the lower-precision arm wins — but there is **no precision-matched hybrid control** here, because no BF16 GGUF of a GDN hybrid is staged.
- The GGUF is a third-party conversion. Its metadata names the same base model, but a different upstream revision cannot be excluded from metadata alone.
- One hybrid model. The three hybrids in #1273 point the same way, but all of them confound quantizer with load path; this is the only one that does not.

**Comment by kekzl** [comment]
## The two checkpoints carry the same weights — proven, not inferred

`linear_attn.norm.weight` is F32 on both sides and reads **bit-identical** between the GGUF and the BF16 SafeTensors (relative difference `0.000000`, layers 0 and 1, range `[+0.6191, +1.0228]`). A different upstream revision would not produce bit-identical norm weights. That closes the "maybe it is a different model" limitation from the issue body: it is the same model, and the 35% gap is imp's.

## Three candidates ruled out

**1. GDN head layout — refuted with the existing switch.**

The per-head GDN vectors really are permuted between the two checkpoints, exactly and identically in every layer:

```
A_log, dt_bias:  GGUF[i] = HF[2i]  for i<16,  GGUF[16+i] = HF[2i+1]
```

That is a de-interleave of the 32 v-heads over 16 k-heads (`linear_num_value_heads=32`, `linear_num_key_heads=16`) — HF stores `[16,2]`, GGUF stores `[2,16]`. `weight_upload.cu` even carries a comment naming a `head_perm(i)` while the code applies only `-exp()`, which reads like a missing transform.

It is not missing. `ModelConfig::gdn_grouped_head_layout` models both layouts and `executor_ssm_gdn.cu` passes it to the scan kernels. A/B on the SafeTensors path with `gdn.layout_override`:

| layout | PPL |
|---|---|
| `grouped` (default for SafeTensors) | **6.6393** |
| `tiled` | 654.2743 |

The default is right and the permutation is already handled. A 100x wrong answer is what the bug would look like; 6.64 is not it.

**2. Norm-weight `1 + W` convention — correct on every norm that needs it.**

Qwen3.5 SafeTensors stores block-norm gammas as deltas. Checked elementwise against the GGUF, layers 0/1/3/7:

| tensor | verdict | `max｜gguf − (hf+1)｜` |
|---|---|---|
| `input_layernorm` | offset needed | 0.0000 |
| `post_attention_layernorm` | offset needed | 0.0000 |
| `self_attn.q_norm` / `k_norm` | offset needed | 0.0000 |
| `linear_attn.norm` | **no offset** | identical without it |

`weight_upload.cu` applies `arch_norm_offset` to exactly the first three families and uploads `ssm_norm_w` without it. Correct on all four.

**3. The `A_log` transform (#1282) — already fixed and confirmed here.** `-exp(HF A_log)` reproduces the GGUF values exactly (same multiset, and identical elementwise once the head permutation above is accounted for).

## Where that leaves it

The remaining difference between the two paths is the large projection weights, which are MXFP4 on the GGUF side and BF16 on the SafeTensors side — the one comparison that cannot be made directly without dequantizing MXFP4. Everything small and high-precision now matches.

Also worth noting for whoever picks this up: the per-layer hidden-state diff is **flat** at 0.17–0.22 across all 32 layers with no step change, which argues against a single mis-loaded tensor family and for something that touches every layer equally.

**Comment by kekzl** [comment]
## A second control removes the confound I had missed

The first control (Qwen3-14B) is dense **and text-only**. The degraded model is hybrid **and multimodal** — its SafeTensors checkpoint carries a vision tower, so the loader turns on M-RoPE (`M-RoPE section [11, 11, 10] (interleaved)`) while the GGUF conversion is text-only and does not. "Hybrid" and "multimodal SafeTensors" were confounded in the comparison, and M-RoPE is not cosmetic — the interleaved and blocked layouts rotate different dimensions by different angles.

Qwen3-VL-4B separates them: dense attention, multimodal, and a GGUF twin is staged.

| model | attention | M-RoPE on the ST path | SafeTensors | GGUF | ordering |
|---|---|---|---|---|---|
| Qwen3-14B | dense | no | **5.3769** (BF16) | 5.4545 (Q6_K) | correct ✅ |
| Qwen3-VL-4B | dense | **yes, `[24,20,20]` interleaved** | **5.6308** (BF16) | 5.7476 (Q4_K_M) | correct ✅ |
| Qwen3.5-4B | **GDN hybrid** | yes, `[11,11,10]` interleaved | 6.6238 (BF16) | **5.0928** (mxfp4) | **inverted ❌** |

`ppl_corpus.txt` for all three rows. The multimodal SafeTensors path with M-RoPE active is correctly ordered, so neither multimodality nor M-RoPE explains the inversion. **GDN is the only thing left that separates row 3 from rows 1 and 2.**

## Two more candidates dead

**cuBLAS FP16-accumulate.** Both paths log `cuBLAS FP16-accumulate prefill: auto → ON`, and the quantized path does not necessarily route its GEMMs through cuBLAS — so the "16-bit" arm could plausibly have been the numerically worse one. It is not:

| arm | `gemm.cublas_fp16_acc=on` | `off` |
|---|---|---|
| BF16 SafeTensors | 6.6186 | 6.6315 |
| mxfp4 GGUF | 5.0930 | 5.0924 |

0.2% and 0.01%. Not a 30% gap.

**The fused Q+gate split.** `attention.gate_concat` is the one switch that sits exactly where #1273 localised the divergence, so it was worth an A/B on both paths:

| arm | `false` (default) | `true` |
|---|---|---|
| BF16 SafeTensors | **6.6238** | 79026.2512 |
| mxfp4 GGUF | **5.0930** | 474573.3890 |

Per-head interleaved is right for both loaders, by four orders of magnitude. The default is not the bug.

**Tokenisation.** Same 13811 tokens on the corpus was not proof the token *ids* match. Checked per token against the embedding-output dump: every token's embedding differs by 0.16–0.33 relative — the magnitude 4-bit quantisation predicts — and **none** by more than 0.5, which is what a different token id would look like. Same ids.

## Running list of what is dead

checkpoint identity · GDN head layout · norm `1+W` offset · `A_log` (#1282) · the large projections after MXFP4 dequant · cuBLAS FP16-accumulate · the fused Q+gate split · tokenisation · multimodality and M-RoPE · run-to-run noise.

**Comment by kekzl** [comment]
## Every GDN knob swept — the gap is invariant to all of them

If the SafeTensors path were taking a different kernel route than the GGUF path, forcing them onto the same route would close the gap. It does not move it at all.

| arm | PPL |
|---|---|
| BF16 default | 6.6392 |
| BF16 `gdn.ref_kernel=true` | 6.6324 |
| BF16 `gdn.chunkwise_scan=false` | 6.6392 |
| BF16 `gdn.fp32_scan=true` | 6.6282 |
| BF16 `gdn.vhead_reorder=true` | **88974.3341** |
| mxfp4 GGUF default | 5.0892 |
| mxfp4 GGUF `gdn.ref_kernel=true` | 5.0963 |

The three scan variants agree within 0.2% on the SafeTensors path — reference, fused and chunkwise all produce the same wrong number, so it is not one kernel's bug. `vhead_reorder` confirms its default: turning it on destroys the model, which is what a *double* reorder looks like on a checkpoint that is already in the layout the scan expects.

## `conv1d` localised, and it settles the identity question for good

The one tensor I had left hanging. Split by channel region (`q | k | v`, 2048 / 2048 / 4096 of the 8192 conv channels):

| region | relative difference |
|---|---|
| q `[0, 2048)` | **0.0000** |
| k `[2048, 4096)` | **0.0000** |
| v `[4096, 8192)` | 1.3390 |

The first differing channel is 4224 = 4096 + 128 — the start of v-head 1, exactly the de-interleave signature already seen on `A_log` and `dt_bias`, and only on the 32 v-heads. q and k, which have 16 heads and therefore no interleave to undo, are **bit-identical between the two checkpoints**. Two files from different producers do not agree to the bit by accident: same weights, confirmed a second way.

## Where this stands

Everything comparable between the two checkpoints matches, every dispatch knob is at its correct setting (each verified by making it *worse* when flipped), and the 30% gap survives all of it. Ruled out so far:

checkpoint identity · GDN head layout · v-head reorder · norm `1+W` offset · `A_log` · `conv1d` · the large projections · cuBLAS FP16-accumulate · the fused Q+gate split · all three GDN scan kernels · FP32 scan · tokenisation · multimodality and M-RoPE · run-to-run noise.

The next technique has to be something other than comparing the two paths to each other — they agree on every input I can reach. A reference forward pass (HF `transformers`) on the same tokens would say which of the two is wrong, and that needs an image with torch, which the imp containers deliberately do not carry.

**Comment by kekzl** [comment]
## Settled: the SafeTensors path is the wrong one

The two imp paths agree on every input I can reach, so comparing them further could not say which is right. A reference forward can. HF `transformers` on the same checkpoint, same corpus, same BF16 precision — pure-torch GDN path (it logs that flash-linear-attention is absent and falls back), 426 weights loaded, the same number imp assigns:

| implementation | precision | PPL | vs reference |
|---|---|---|---|
| **HF transformers** | BF16 | **4.6990** | — |
| imp, GGUF mxfp4 | 4-bit | 5.0928 | **+8.4%** |
| imp, SafeTensors | **BF16** | 6.6238 | **+41.0%** |

imp's GGUF path costs 8.4% over the BF16 reference, which is what 4-bit quantisation of a 4B model costs — that number also cross-checks the perplexity definition, since a mismatched convention would not land that close. imp's SafeTensors path costs **41%, at the reference's own precision**.

So this is not "the GGUF is suspiciously good". **The SafeTensors hybrid path is wrong**, by an independent implementation, and the GGUF path is approximately right.

`ppl_corpus.txt`; HF tokenises it to 200 tokens against imp's 192–199, a difference far too small to matter at this magnitude.

## What that means for #1273

Every hybrid NVFP4 checkpoint in #1273 is read through this path. Its healthy twins are all GGUF, read through the path that measures correct here. The degradation attributed there to NVFP4 quantising the gated `q_proj` is at minimum confounded with a ~40% engine error that has nothing to do with quantisation — the same error is present at BF16, where there is no quantisation at all.

Reproduce with:

```
docker run --rm --gpus all -v /mnt/data/models:/dmodels:ro -v <repo>/tools/analysis:/corp:ro \
  pytorch/pytorch:2.9.1-cuda12.8-cudnn9-runtime \
  bash -lc "pip -q install -U 'transformers>=4.60' accelerate && python /sp/ref_ppl.py /dmodels/Qwen3.5-4B-BF16 /corp/ppl_corpus.txt"
```

Next step is a layer-by-layer diff against that reference rather than against the GGUF — the reference can say *where* the two diverge, which the GGUF comparison could not because 4-bit noise swamps it.

**Comment by kekzl** [comment]
Fixed in #1289 (`ebf39fb0`), and it closes #1273 with it.

The per-layer RMS profile against the HF reference is what pointed at it: imp's SafeTensors path matched HF **within 0.4% across all 32 layers** (ratio 0.987–1.014) — closer than the GGUF path, which sits at 0.94 because its weights are 4-bit — while its perplexity was 41% off. States right, output wrong, so the error had to be after the last layer.

From there it was one comparison. `model.language_model.norm.weight` needs the `1 + W` offset (`max|gguf − (hf + 1)| = 0.0000`), and it was the one norm not getting it: `weight_upload.cu` uploaded `out_norm` through `upload_unquantized_weight`, which has no offset parameter, while every other norm went through `upload_weight(..., ctx.arch_norm_offset, ...)`.

| implementation | precision | before | after |
|---|---|---|---|
| HF transformers | BF16 | 4.6990 | — |
| imp SafeTensors | BF16 | 6.6238 | **4.6621** |
| imp GGUF mxfp4 | 4-bit | 5.0928 | 5.0930 (untouched) |

On `ppl_corpus_45k` the ordering is finally right too: BF16 **12.6794 → 9.2624**, now better than its own 4-bit twin at 9.3793 instead of 35% worse.

## Note on the fourteen dead candidates

Ruling them out — each by making its switch *worse* rather than by inspection (654 / 79026 / 88974 / 474573) — emptied the format axis and made it clear the fault was not there. But none of them could have found this one: the missing line had no switch to flip. That took the reference implementation.
